frontend/src/stores/useProjectStore.ts

```diff
@@ -224,6 +245,94 @@ const useProjectStore = create<ProjectStore>()(persist((set: SetState<ProjectSto
+  })),
+
+  insertGroup: (createData, isTemplate = false) => {
+    // Check that we are inserting into same project type
+    if (createData.projectType !== get().project.projectType) {
+      return { type: InsertGroupResponseType.FAIL, body: `You cannot insert elements from a ${createData.projectType} project into a ${get().project.projectType} project.` }
+    }
+    // Check that for transitions being inserted, to and from states are also inserted
+    let missingState = false
+    createData.transitions.forEach((transition) => {
+      let toFound = false
+      let fromFound = false
+      createData.states.forEach((state) => {
+        if (state.id === transition.from) {
+          fromFound = true
+        }
+        if (state.id === transition.to) {
+          toFound = true
+        }
+      })
+      // If a transition has to or from missing, flag to return error
+      if (!(toFound && fromFound)) {
+        missingState = true
+      }
+    })
+    if (missingState) {
+      return { type: InsertGroupResponseType.FAIL, body: 'Sorry, there was an error.' }
+    }
+    set(produce(({ project }: { project }) => {
+      let isInitialStateUpdated = false
+      const isNewProject = createData.projectSource !== project._id
+      const positionOffset = isTemplate ? 0 : PASTE_POSITION_OFFSET
+      const newTransitions = structuredClone(createData.transitions)
+      newTransitions.forEach(transition => {
+        transition.from = null
+        transition.to = null
+      })
+      const nextStateId = nextIDFor(project.states)
+      createData.states.forEach((state, i) => {
+        // TODO: ensure position isn't out of window
+        // Probably will have to take adjusting position out of this function
+        [state.x, state.y] = [state.x + positionOffset, state.y + positionOffset]
+        const newId = nextStateId + i
+        // Update transitions to new state id
+        createData.transitions.forEach((transition, i) => {
+          if (transition.from === state.id && newTransitions[i].from === null) {
+            newTransitions[i].from = newId
+          }
+          if (transition.to === state.id && newTransitions[i].to === null) {
+            newTransitions[i].to = newId
+          }
+        })
+        // Update initial state id if applicable
+        if (createData.initialStateId === state.id && !isInitialStateUpdated) {
+          createData.initialStateId = newId
+          isInitialStateUpdated = true
+        }
+        state.id = newId
+        // createState
+        project.states.push({ ...state })
+      })
+      createData.transitions = newTransitions
+      const nextCommentId = nextIDFor(project.comments)
+      createData.comments.forEach((comment, i) => {
+        // TODO: ensure position isn't out of window
+        [comment.x, comment.y] = [comment.x + positionOffset, comment.y + positionOffset]
+        const newId = nextCommentId + i
+        comment.id = newId
```

Comment:
```suggestion
        comment.id = nextCommentId + i
```
Same small nitpick. Good in places like this though since it means one less variable needs to be kept in mental context when reading code